### PR TITLE
Shopping cart model

### PR DESCRIPTION
**Pull request by evgenkolner**

### CODE REVIEW

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+@Tag(name = "Shopping cart management", description = "Shopping cart management")
+@RequiredArgsConstructor
+@RestController
+@Validated
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PreAuthorize("hasRole('USER')")
```
Comment (teract10s) [inline src/main/java/bookstore/controller/ShoppingCartController.java:35]: is it work without prefix "ROLE_"?

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    @Operation(summary = "Add book to cart",
+            description = "Add book to cart")
+    public ShoppingCartDto addBookToCart(Authentication authentication,
+                                         @RequestBody CartItemRequestDto requestDto) {
+        User user = (User) authentication.getPrincipal();
```
Comment (teract10s) [inline src/main/java/bookstore/controller/ShoppingCartController.java:50]: Is it controller layer behavior? I had comment from mentors that here would be better to move in service layer

File: src/main/java/bookstore/dto/cartitem/CartItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package bookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.Positive;
+
+public record CartItemRequestDto(
+        @Positive
+        Long bookId,
+
+        @Min(1)
```
Comment (teract10s) [inline src/main/java/bookstore/dto/cartitem/CartItemRequestDto.java]: You may also replace it with positive and add null check where it has to be

File: src/main/java/bookstore/dto/shoppingcart/ShoppingCartRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package bookstore.dto.shoppingcart;
+
+public record ShoppingCartRequestDto(
+        int quantity) {
```
Comment (teract10s) [inline src/main/java/bookstore/dto/shoppingcart/ShoppingCartRequestDto.java]: add checks here

**Review by teract10s — CHANGES_REQUESTED** [review]
Good job. I left some minor comments

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        return shoppingCartService.findById(user.getId());
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    @Operation(summary = "Add book to cart",
+            description = "Add book to cart")
+    public ShoppingCartDto addBookToCart(Authentication authentication,
+                                         @RequestBody CartItemRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/ShoppingCartController.java]:
```suggestion
                                         @RequestBody @Valid CartItemRequestDto requestDto) {
```

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        return shoppingCartService.addToCart(user.getId(), requestDto);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PutMapping(("/cart-items/{id}"))
+    @Operation(summary = "Update cart item",
+            description = "Update cart item")
+    public ShoppingCartDto updateCartItem(Authentication authentication,
+                                          @PathVariable @Positive Long id,
+                                          @RequestBody ShoppingCartRequestDto quantity) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/ShoppingCartController.java]:
```suggestion
                                          @RequestBody @Valid ShoppingCartRequestDto quantity) {
```

File: src/main/java/bookstore/controller/ShoppingCartController.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+@Tag(name = "Shopping cart management", description = "Shopping cart management")
+@RequiredArgsConstructor
+@RestController
+@Validated
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @PreAuthorize("hasRole('USER')")
```
It works without ROLE
Check all places

File: src/main/java/bookstore/dto/cartitem/CartItemRequestDto.java (hunk 3 above)
Comment (Elena-Bruyako) [inline src/main/java/bookstore/dto/cartitem/CartItemRequestDto.java]: Positive is better

File: src/main/java/bookstore/mapper/CartItemMapper.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.NullValueCheckStrategy;
+
+@Mapper(componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl",
+        uses = BookMapper.class)
+
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/mapper/CartItemMapper.java]: remove empty line

File: src/main/java/bookstore/mapper/ShoppingCartMapper.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.NullValueCheckStrategy;
+
+@Mapper(componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl",
+        uses = CartItemMapper.class)
+
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/mapper/ShoppingCartMapper.java]: the same

File: src/main/java/bookstore/service/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final UserRepository userRepository;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+
+    @Transactional(readOnly = true)
+    @Override
+    public ShoppingCartDto findById(Long id) {
+        return shoppingCartMapper.toDto(shoppingCartRepository.findById(id).orElseThrow(
+                () -> new RuntimeException("Cannot find shopping cart")));
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/service/ShoppingCartServiceImpl.java]: RuntimeException is too general, please use custom one

File: src/main/java/bookstore/service/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        CartItem cartItem = new CartItem();
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItem.setQuantity(requestDto.quantity());
+        shoppingCart.getCartItems().add(cartItem);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    private ShoppingCart createShoppingCartForUser(User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/service/ShoppingCartServiceImpl.java]: private methods should be after all public methods

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! let's improve your solution!

**Review by Elena-Bruyako — APPROVED** [review]
Good job!

**Pull request merged**